### PR TITLE
Transformer migrator

### PR DESCRIPTION
**Pull request by marcvberg**
#### Description of changes
Adds in an auto-migrator that will run when schema is transformed, this can be triggered on a `amplify api gql-compile` or an `amplify push`. It will update the feature flag to the V2 transformer and it will write any old schema files to the same path but with a .bkp extension

As this is meant for migrating existing projects, it won't work if you initialize an amplify project and try to migrate before pushing your API.

#### Description of how you validated changes
Executed manually

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-transformer-migrator/src/__tests__/migration/migrate-tests.ts
Comment (edwardfoyle) [inline packages/amplify-graphql-transformer-migrator/src/__tests__/migration/migrate-tests.ts:74]: I think for most of these tests we can use [`test.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) to be more succinct

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:12]: You can put `teamProviderInfo[env]` in a variable since it's used below as well. And you don't need the extra `&&` check if you use optional chaining.

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:14]: You don't need all of these checks if you use optional chaining.

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:38]: I believe `readdirSync()` will return an empty array if the directory is empty, so you don't need the `files &&` part of the check below.

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:47]: You don't need to specify the type on both sides of the equal sign:

```
let unsupportedDirSet = new Set<string>();
```

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:48]: This file also looks like the linter did not run.

File: packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts:31]: Just pointing out that we need a real value for this.

File: packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts:17]: I think the preferred style in our codebase is:

```
import * as path from 'path';
```

File: packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts:153]: Is it possible for this to throw? If so, it should be wrapped in its own `try...catch`. If it throws, we may want to surface that error as well.

File: packages/amplify-graphql-transformer-migrator/src/utils.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/utils.ts:74]: Please use `path.join()` instead of hard coding `/`

File: packages/amplify-graphql-transformer-migrator/src/utils.ts
Comment (cjihrig) [inline packages/amplify-graphql-transformer-migrator/src/utils.ts:100]: Can this just be a `filter()` call?

**Review by cjihrig — COMMENTED** [review]
Left some comments. This will likely require at least one more review once the comments are addressed.

File: packages/amplify-graphql-transformer-migrator/src/__tests__/migration/migrate-tests.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/__tests__/migration/migrate-tests.ts:74]: Planning to make changes to the testing but I think we're going to ship this unhooked to the CLI flows first to unblock other testing

File: yarn.lock
Comment (marcvberg) [reply yarn.lock:12999]: no idea, I'll just remove it

File: packages/amplify-graphql-transformer-migrator/src/constants/graphql-directives.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/constants/graphql-directives.ts:1]: Attila doesn't want any dependencies on V1 packages because it makes it easier for us to move this out of the CLI. I have more changes to make like renaming the file to be V1 specific since we're going to just try and pre-ship it unhooked. I think that should be fine since we're not focused on feature addition in V1

File: packages/amplify-graphql-transformer-migrator/src/migrators/auth/fieldAuth.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/auth/fieldAuth.ts:4]: I'm still not clear how the linter is missing this or what I need to configure, but I'll take another pass to check for semicolons

File: packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/schema-migrator.ts:31]: Yeah this is going to be a banner message, I'll get it fixed before the actual ship

File: packages/amplify-graphql-transformer-migrator/src/migrators/auth/index.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/auth/index.ts:16]: This is going to be part of the changes done this week

File: packages/amplify-graphql-transformer-migrator/src/migrators/connection/index.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/connection/index.ts:29]: I can revisit this when I'm rewriting the migrator code, many of these functions Rene wrote are only compatible with each other and mixing it with the common code breaks. During the rewrite I can get away from these

File: packages/amplify-graphql-transformer-migrator/src/migrators/connection/index.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/connection/index.ts:47]: To be changed before actual ship

File: packages/amplify-graphql-transformer-migrator/src/migrators/generators/index.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/generators/index.ts:1]: I know the linter is running in the new package it just seems to be a lot less selective for some reason, I'll note this down as a todo item

File: packages/amplify-graphql-transformer-migrator/src/migrators/generators/index.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/migrators/generators/index.ts:4]: Yeah and I changed it but I had to change back, the code in here isn't 100% compatible with transformer-common so I'll fix it with the rewrite this week

File: packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts
Comment (marcvberg) [reply packages/amplify-graphql-transformer-migrator/src/schema-inspector.ts:12]: I didn't know how to do optional chaining with brackets - I'll fix that

**Comment by cjihrig** [comment]
If we want to fast track this and have it disconnected from the rest of the CLI, the only files changed in this PR should be:

- Things in `packages/amplify-graphql-transformer-migrator`
- `jest.config.js`
- `yarn.lock`

**Review by cjihrig — APPROVED** [review]
LGTM, pending CI, based on the conclusion from today's meeting (to merge this as is with no connection to the rest of the CLI and iterate on it).

**Pull request merged**
